design/search.md

```diff
@@ -113,4 +115,31 @@ Permissions:
+
+### Search & query-translation
+The search service in general operates by taking in queries and translating them to SQL, fetching results, and then formatting them for consumption. The return formats will be similar to our existing APIs, so I won't go into detail on those, but the query translation bears some additional detail.
+
+I'll use a shorthand to express the intent of the query, though it will be done in the actual service with a more verbose JSON-based format, so as to avoid spending too much time writing parsers for a custom format.
+
+First, a simple example. Imagine a query like `label:"*example.bam*" AND path_prefix:"/iplant/home/atreyu/"`, which is about what the current simple search does. That query is expressed as two clauses, one for the label and one for the path prefix, joined with AND. Elasticsearch, and our current search service, use a structure with three lists: ES calls them must, should, and must_not; our search service uses all, any, none; so, in that option, the two clauses would appear in the 'must' or 'all' list. Upon getting the query, the service would generate fragments of SQL from each, intended to be combined into a WHERE clause. In this case, that would be sufficient, though in other cases the service may need to be able to modify other pieces of a SELECT query such as joins, window functions, or CTEs. For the label clause, assuming our defaults expect case-insensitive search with wildcards on both sides, the fragment would be something like `label ILIKE '%example.bam%'`; for the path, case sensitivity would need to be turn on, so it would be `path LIKE '/iplant/home/atreyu/%'`. (Despite the leading wildcard, a trigram index can help with the label clause). The produced query would therefore look roughly like: `SELECT (...relevant columns to produce JSON output) FROM data (... any joins needed to produce the output) WHERE label ILIKE '%example.bam%' AND path LIKE '/iplant/home/atreyu/%';`.
+
+For a more complex example (but less intermediate explanation): `any(label:"*example.bam*", label:"*test.bam*"), all(metadata:(attribute:"ipc-filetype",value:"unknown"), metadata(attribute:"usecase-specific",value:"bam*")` would create fragments for all the clauses, combine with AND and OR as needed by any/all. Therefore in something like pseudocode:
+
+```
+"(" + [labelClause1, labelClause2].join(" OR ") + ") AND " + [metadataClause1, metadataClause2].join(" AND ")
+```
+
+The metadata clauses would need to create fragments like `EXISTS (SELECT 1 FROM metadata WHERE (... stuff that ties it to the external query's data IDs) AND metadata.attribute = "..." AND metadata.value = "...")`; or, depending on what goes in the initial part of the query (what columns are selected how), it might be possible to use array functions and matching, or json operators, to avoid the subselect.
+
 ## Estimates
+// test performance, compare
```

Comment:
No idea what kind of performance before-and-after kind of data would be good to know (other than a high level thing such as a search with 1000 results can be returned in X seconds in the old vs. new implementation) or how to do that without just implementing things, but would be nice to see if possible.

Reply:
I was figuring what I'd propose here was a set of somewhat representative cases -- simple search for both users with a little and a lot of data, data commons type searches, and probably a few complex things (especially involving metadata), plus similar for apps/analyses and try to run them against versions of both underlying systems -- so we wouldn't need to write all the indexing and translation tools, just enough to have a static dataset to run against that looks like real data.